Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +21,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(currentNew_ < 0)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time;
+    double totalHours;
+    int hours;
+    int remainingHours;
+    int minutes;
+    int seconds;
+
+    if(isCharging())
+        totalHours = qAbs(totalAmpHoursUsed() / currentNew_);
+    else
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / currentNew_;
+
+    hours = (int)totalHours;
+    minutes = (int)((totalHours - hours) * HMS_CONVERSION_FACTOR);
+    seconds = (int)((totalHours * HMS_CONVERSION_FACTOR - hours * HMS_CONVERSION_FACTOR - minutes) * HMS_CONVERSION_FACTOR);
+
+    //uses msec as a place holder for the remaining hours as the hours in QTime could only hold up to 24
+    remainingHours = ((int)(totalHours/24)) * 24;
+    hours -= remainingHours;
+
+    time.setHMS(hours, minutes, seconds, remainingHours);
+
+    return time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(timeOld.isNull()) //checks if timeOld has no a value (first itteration)
```

Comment:
`no a value` is not english, also misspelled iteration.

This kinda just reiterates the line same line, explain why you are checking this.